### PR TITLE
Feature/beta-viz

### PR DESCRIPTION
**Pull request by chrismbryant**
This is a first attempt at producing beta distribution visualizations to go with each product review. I've removed the 95% confidence interval calculations from the display for now until we decide what user experience we want to have. To make the visualization, I ended up using D3, which really increased the size of the build (maybe there's a way to explicitly import just the stuff I use like with the stdlib imports?). Also, right now, the SVGs that hold the visualization are statically sized, but we probably would want them to dynamically resize with the div that contains them.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,4 +1,5 @@
 **/.DS_Store
 .idea/
 dist/
-node_modules/
\ No newline at end of file
+node_modules/
+package-lock.json
```
Comment (aeciorc) [inline .gitignore]: Don't ignore this, it can cause some weird bugs

File: src/shared/betaviz.js [hunk 2]
```diff
@@ -0,0 +1,403 @@
+import calculations from "./calculations";
+import * as d3 from "d3";
```
Comment (aeciorc) [inline src/shared/betaviz.js]: It seems like the only thing keeping you from being able to selectively import only the stuff you need is that currently you access the cmap dynamically from d3. If you do need that, a solution would be to import `interpolateSpectral` and others you might use and group them on a colorsMap object.

**Comment by aeciorc** [comment]
This looks nice. I don't know enough about data science to properly review this, so I just left a suggestion for the bundle size issue.

File: .gitignore (hunk 1 above)
Comment (chrismbryant) [reply .gitignore]: Do we have to commit `package-lock.json` then?

Comment (aeciorc) [reply .gitignore]: We do, at least npm recommends that. It keeps tracks of the versions of subdependencies and the order of installation

**Comment by chrismbryant** [comment]
@aeciorc Thanks for the suggestion; I tried doing that, importing just what I needed and putting some color maps into an object, but it doesn't seem to have changed the bundle size (still about 4.5 Mb)...not sure if I'm missing something

File: .gitignore (hunk 1 above)
Comment (chrismbryant) [reply .gitignore]: Noted. I'll commit it then

**Comment by aeciorc** [comment]
> @aeciorc Thanks for the suggestion; I tried doing that, importing just what I needed and putting some color maps into an object, but it doesn't seem to have changed the bundle size (still about 4.5 Mb)...not sure if I'm missing something

Ok, will investigate this tomorrow

**Comment by musicin3d** [comment]
On first pass...

Having never taken a course on statistics or numerical analysis, I don't know what to do with the graph. But if it helps you do your thing then coolio. Where can I learn about it?

I would suggest that we use `let` instead of `var`. It is a newer language feature that has more straightforward scoping and initialization. It was introduced at the same time as `const`.

> let allows you to declare variables that are limited to a scope of a block statement, or expression on which it is used, unlike the var keyword, which defines a variable globally, or locally to an entire function regardless of block scope. The other difference between var and let is that the latter is initialized to a value only when a parser evaluates it (see below).
> 
> Just like const the let does not create properties of the window object when declared globally (in the top-most scope).
> 
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/let

File: src/shared/betaviz.js [hunk 3]
```diff
@@ -0,0 +1,416 @@
...
+import {select as d3select} from "d3";
+import {range as d3range} from "d3";
+import {scaleLinear as d3scaleLinear} from "d3";
+import {curveLinear as d3curveLinear} from "d3";
+import {max as d3max} from "d3";
+import {transpose as d3transpose} from "d3";
+import {area as d3area} from "d3";
+import {interpolateSpectral as d3interpolateSpectral} from "d3";
+import {interpolateRdYlBu as d3interpolateRdYlBu} from "d3";
+import {interpolateRdYlGn as d3interpolateRdYlGn} from "d3";
```
Comment (aeciorc) [inline src/shared/betaviz.js]:
```suggestion
import { select as d3select } from "d3-selection";
import {
  range as d3range,
  transpose as d3transpose,
  max as d3max
} from "d3-array";
import { scaleLinear as d3scaleLinear } from "d3-scale";
import { curveLinear as d3curveLinear, area as d3area } from "d3-shape";
import {
  interpolateSpectral as d3interpolateSpectral,
  interpolateRdYlBu as d3interpolateRdYlBu,
  interpolateRdYlGn as d3interpolateRdYlGn
} from "d3-scale-chromatic";
```

File: package.json [hunk 4]
```diff
@@ -18,6 +18,9 @@
     "webpack-cli": "^3.3.11"
   },
   "dependencies": {
-    "@stdlib/stdlib": "^0.0.91"
+    "@stdlib/stdlib": "^0.0.91",
+    "d3": "^5.15.0",
```
Comment (aeciorc) [inline package.json]:
```suggestion
    "d3-array": "^2.4.0",
    "d3-scale": "^3.2.1",
    "d3-scale-chromatic": "^1.5.0",
    "d3-selection": "^1.4.1",
    "d3-shape": "^1.3.7",
```

**Comment by aeciorc** [comment]
> > @aeciorc Thanks for the suggestion; I tried doing that, importing just what I needed and putting some color maps into an object, but it doesn't seem to have changed the bundle size (still about 4.5 Mb)...not sure if I'm missing something
> 
> Ok, will investigate this tomorrow

@chrismbryant  Check my suggestions. It should decrease the size of inject.js to 1.77 Mb

File: src/shared/calculations.js [hunk 5]
```diff
@@ -9,46 +10,132 @@ import quantile from "@stdlib/stats/base/dists/beta/quantile";
...
+     * from the Amazon rating percentage histogram).
+     *
+     * @param {Object.<string, (number|number[])>} ratingData - object containing
+     *     the average rating under key "avg" and/or the 5-element rating
+     *     distribution (fraction of ratings at each star value) under key "dist".
+     * @param {string} method - which method to use to compute proportion of
+     *     reviews which were positive; must be either "avg" or "dist".
+     * @returns {number} proportion - proportion of ratings which were positive.
+     */
+    getProportionPositive(ratingData, method = "avg") {
```
Comment (musicin3d) [inline src/shared/calculations.js:56]: I'm curious why this was introduced as an alternative to directly calling the underlying methods.

File: src/shared/calculations.js [hunk 6]
```diff
@@ -9,46 +10,132 @@ import quantile from "@stdlib/stats/base/dists/beta/quantile";
...
+     *     distribution (fraction of ratings at each star value) under key "dist".
+     * @param {number} numRatings - total number of ratings for this product.
+     * @param {string} method - which method to use to compute proportion of
+     *     reviews which were positive; must be either "avg" or "dist".
+     * @returns {Object.<string, number>} betaParams - object containing "alpha"
+     *     and "beta" parameters for the beta distribution, as well as the
+     *     estimated "proportion" and number "numPositive" of ratings which were 
+     *     positve, and the total number of ratings "numRatings".
+     */
+    getBetaParams(ratingData, numRatings, method = "avg") {
```
Comment (musicin3d) [inline src/shared/calculations.js]: (I just wrote a beautifully composed comment and lost it when I clicked a link. :cry: Here is but a memory of what was...)

At this point, we're accepting a parameter that is only used further down the call stack (only to determine which method to drive). There's also a bit of redundancy as the user has to specify "avg" or "dist" twice. I can't help but think there's a better way to get where we're going.

Perhaps we should just have two separate methods, like`getBetaParamsDist(...)` and `getBetaParamsAvg(...)`?

The [rule of three](https://en.wikipedia.org/wiki/Rule_of_three_(computer_programming)) would suggest that it's too early to try to DRY this anyway.

File: src/shared/calculations.js [hunk 7]
```diff
@@ -9,46 +10,132 @@ import quantile from "@stdlib/stats/base/dists/beta/quantile";
...
+     *     distribution (fraction of ratings at each star value) under key "dist".
+     * @param {number} numRatings - total number of ratings for this product.
+     * @param {string} method - which method to use to compute proportion of
+     *     reviews which were positive; must be either "avg" or "dist".
      * @returns {ConfidenceInterval}
      */
-    evaluateAverageRating(avgRating, numRatings) {
-        const avgProb = (avgRating - 1) / 4;
-        const numPositive = Math.floor(avgProb * numRatings);
+    evaluateRatings(ratingData, numRatings, method = "avg") {
```
Comment (musicin3d) [inline src/shared/calculations.js]: Same comment as above.

File: src/shared/calculations.js (hunk 6 above)
Comment (musicin3d) [reply src/shared/calculations.js]: If we really feel like it's necessary to have different "modes" then perhaps we should add a `mode` property.

**Comment by musicin3d** [comment]
Disclaimer: I'm a loud mouth in code review. I do try to present things open handedly and open mindedly. But I figure it's better to risk a little discomfort than to wish you'd spoken up later on. Feel free to tell me to can it if I overstep. :laughing:

File: src/shared/betaviz.js (hunk 2 above)
Comment (musicin3d) [reply src/shared/betaviz.js]: Looks like this suggestion was applied. Resolving

**Comment by chrismbryant** [comment]
@aeciorc Thanks for the suggestions!

@musicin3d Thanks, I changed all `var`s to `let`s. And no worries, please continue giving feedback whenever something can be done better! Regarding the visualizations, I'm using the number of positive and negative ratings to get a probability distribution of the underlying "positive experience rate". This is what @3b1b started to get at in his first video, and what I assume he'll address more in the 2nd and especially 3rd video of his "probabilities of probabilities" series. The [blog post](https://www.johndcook.com/blog/2011/09/27/bayesian-amazon/) by John Cook he mentioned is also a good start for understanding what's going on.

Originally, I was only planning on testing out the visualization capabilities by plotting a number line with a rectangle encompassing the 95% confidence interval, but I realized that I might as well go all the way and show the true distribution where that confidence interval came from (the "95%" there means that 95% of the area inside the distribution is contained within that interval, with 2.5% hanging out to the left, and 2.5% hanging out to the right). The more data we have, the skinnier that distribution is, and the higher the positive experience rate is, the farther to the right its peak is pushed. So the best product has a skinny distribution pushed all the way to the right. My D3 visualizations show the distribution from 0% to 100% with a line at the 50% mark and dots at every 10% increment in between.

**Review by musicin3d — COMMENTED** [review]
Without knowing the math principles, this all looks reasonable. 

I did leave a comment about the interface of calculations.js. It think it's important to set good precedents early on, so I'm withholding explicit approval until we can discuss those points.

I see no obvious issues in betaviz.js and it seems to accomplish your goals effectively. :+1:

**Comment by musicin3d** [comment]
I like this naming convention for feature branches. I'll start using it on the next PR.

**Review by aeciorc — APPROVED** [review]

File: src/shared/calculations.js (hunk 6 above)
Comment (chrismbryant) [reply src/shared/calculations.js]: I get what you're saying; it does feel redundant, but I'm not exactly sure how to fix it. The idea was that `ratingData` would be a dictionary of all the possible forms of rating data you care about, like:
```
ratingData = {
  "dist": [0.1, 0, 0, 0.4, 0.5],
  "avg": 4.7
}
```
So you can just specify that once and decide later whether you want to use "avg" or "dist" to generate your statistics. Now that I say that though, if that's what I'm going for, it seems like I might also want to include `numRatings` inside of that dictionary...

Not sure where to go from here. Thoughts?

**Comment by aeciorc** [comment]
I feel like you’re trying to optimize this a little too early. I would do away with the ratingData object and have a method for getting the beta using average and another for dist, both taking only the params they use. Then getaBetaParams would take the proportion itself rather than the method
--Edit
I replied this via email and didn't realize it would show up here, I meant to rely your latest comment

**Comment by musicin3d** [comment]
> I replied this via email and didn't realize it would show up here, I meant to rely your latest comment

This one? https://github.com/chrismbryant/amazon-confidence-interval/pull/19#discussion_r402028060

**Comment by musicin3d** [comment]
Random feedback on the feature in general: I unwittingly had this active while shopping for computer parts. It was very useful. :)

File: src/shared/calculations.js (hunk 6 above)
Comment (musicin3d) [reply src/shared/calculations.js]: I'm a huge believer in taking the dumb approach first and iterating from there as real-world problems arise. Would you be opposed to having an "average" and a "distribution" version of each of those functions, for now?

**Comment by aeciorc** [comment]
> > I replied this via email and didn't realize it would show up here, I meant to rely your latest comment
> 
> This one? [#19 (comment)](https://github.com/chrismbryant/amazon-confidence-interval/pull/19#discussion_r402028060)

Yes, that one

**Comment by chrismbryant** [comment]
@aeciorc and @musicin3d , thanks for the feedback. The suggestion to make the functions take `proportion` as an argument made things clearer for me. Does the code look better now?

**Review by musicin3d — APPROVED** [review]
This all looks good to me!

**Comment by musicin3d** [comment]
Thoughts on whether we should delete branches after merge?

**Comment by chrismbryant** [comment]
@musicin3d At the moment, I don't have strong feelings either way. I think it can be up to the person who made the branch. If the branch is a self-contained feature that you don't plan on re-visiting, then delete the branch after merge. But if the branch represents a larger feature that is being done in multiple steps, then leave it open.

**Pull request merged**